testsuite/tests/basic/patmatch_for_multiple.ml

```diff
@@ -35,25 +25,179 @@ match (3, 2, 1) with
 | _ -> false
 ;;
 [%%expect{|
-(let
-  (*match*/96 = 3
-   *match*/97 = 2
-   *match*/98 = 1
-   *match*/99 = (makeblock 0 *match*/96 *match*/97 *match*/98))
+(let (*match*/96 = 3 *match*/97 = 2 *match*/98 = 1)
   (catch
-    (catch
-      (let (*match*/100 =a (field 0 *match*/99))
-        (catch
-          (let (*match*/101 =a (field 1 *match*/99))
-            (if (!= *match*/101 3) (exit 7)
-              (let (*match*/102 =a (field 2 *match*/99)) (exit 5 *match*/99))))
-         with (7)
-          (if (!= *match*/100 1) (exit 6)
-            (let
-              (*match*/104 =a (field 2 *match*/99)
-               *match*/103 =a (field 1 *match*/99))
-              (exit 5 *match*/99)))))
-     with (6) 0)
-   with (5 x/94) (seq (ignore x/94) 1)))
+    (if (!= *match*/97 3)
+      (if (!= *match*/96 1) 0
+        (exit 4 (makeblock 0 *match*/96 *match*/97 *match*/98)))
+      (exit 4 (makeblock 0 *match*/96 *match*/97 *match*/98)))
+   with (4 x/94) (seq (ignore x/94) 1)))
 - : bool = false
 |}];;
+
+(* Regression test for #3780 *)
+let f a b =
+  match a, b with
+  | ((true, _) as _g)
+  | ((false, _) as _g) -> ()
+[%%expect{|
+(let (f/106 = (function a/108 b/109 0))
+  (apply (field 1 (global Toploop!)) "f" f/106))
+val f : bool -> 'a -> unit = <fun>
+|}];;
```

Comment:
There would be less noise in the expect-block if you removed the binding (i.e. if you had `(fun a b -> match a, b with ...)`).
That's also true for all the later tests in this file.